src/host.rs

```diff
@@ -24,9 +24,13 @@ pub(crate) enum HostInternal {
     Ipv6(Ipv6Addr),
 }
 
-impl<S> From<Host<S>> for HostInternal {
+impl<S> From<Host<S>> for HostInternal
+where
+    S: ToString,
+{
     fn from(host: Host<S>) -> HostInternal {
         match host {
+            Host::Domain(ref s) if s.to_string().is_empty() => HostInternal::None,
```

Comment:
What exactly is this doing? This allocates a new string every time. Do we really need the generics here? Couldn't it be `From<Host<String>>` or `From<Host<&'_ str>>`?

Reply:
We actually don't need the generics here.
The main goal of this change is to make sure HostInternal::None is returned when an empty domain is given.

Reply:
How can it be empty?

Reply:
Here's a list of the failing tests that occur if I don't add this check: 

```
failures:
    "file://hi/x".host = "" 
    "file://hi/x".hostname = "" 
    "file://y/".host = "loc%41lhost" 
    "file://y/".hostname = "loc%41lhost" 
```
Maybe the fix should occur somewhere else, if after the host parsing ?

```
........................................................................................................................................................................thread '"file://hi/x".hostname = "" ' panicked at 'called `Result::unwrap()` on an `Err` value: "!( !host_str.is_empty() ) for URL \"file:///x\""', src/libcore/result.rs.:999:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
..thread '"file://hi/x".host = "" ' panicked at 'called `Result::unwrap()` on an `Err` value: "!( !host_str.is_empty() ) for URL \"file:///x\""', src/libcore/result.rs:999:5
.......F.F...thread '"file://y/".host = "loc%41lhost" ' panicked at 'called `Result::unwrap()` on an `Err` value: "!( !host_str.is_empty() ) for URL \"file:///\""thread '', "file://y/".hostname = "loc%41lhost" src/libcore/result.rs' panicked at ':called `Result::unwrap()` on an `Err` value: "!( !host_str.is_empty() ) for URL \"file:///\""999', :src/libcore/result.rs5:
999:5
..F.F............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
```

Reply:
Where is the `<From<Host<S>>>::from` call done in those?

Reply:
I m not really sure this is the question you're asking me, but the conversion I'm trying to work on is [host.into()](https://github.com/servo/rust-url/blob/master/src/lib.rs#L1677)